### PR TITLE
[DH-496] Allow filtering by company name - sector - country - region.

### PR DESCRIPTION
**Pull request by elcct**
This brings `Contact` filters in line with requirements, which state we need 4 filters:

- Company Name (with partial matching)
- Sector
- Country
- UK Region

### CODE REVIEW

File: datahub/search/contact/test/test_views.py [hunk 1]
```diff
@@ -27,15 +27,49 @@ def test_search_contact(self, setup_es, setup_data):
...
+
+    def test_filter_contact(self, setup_es, setup_data):
+        """Tests matching contact using multiple filters."""
+        contact = ContactFactory(address_same_as_company=True)
+        company = contact.company
+        company.name = 'SlothsCats'
+        company.trading_address_country_id = Country.united_kingdom.value.id
+        company.uk_region_id = UKRegion.east_of_england.value.id
+        company.sector_id = Sector.renewable_energy_wind.value.id
+        company.save()
```
Comment (reupen) [inline datahub/search/contact/test/test_views.py]: Could create a company first and then pass it to ContactFactory()?

**Comment by reupen** [comment]
Would be nice to cover a few more scenarios in the tests (e.g. correct handling of companies/contacts without a uk_region, testing of cases where filters don't match results).

File: datahub/search/contact/test/test_views.py (hunk 1 above)
Comment (elcct) [inline datahub/search/contact/test/test_views.py]: I see no reason why not. Changed.

**Comment by elcct** [comment]
I will add some more tests.

**Comment by elcct** [comment]
@reupen @marcofucci I added a little more tests for `Contact` filters and also cleaned up the search views.

**Comment by reupen** [comment]
I think it would be good to split off the search view refactor to a separate PR. For the interactions search app I was looking at using DRF serialisers:
https://github.com/uktrade/data-hub-leeloo/compare/feature/DH-488-interactions-in-search?expand=1

**Comment by elcct** [comment]
That's nice. I can definitely split that and then also incorporate DRF serializers.

**Comment by marcofucci** [comment]
Yes maybe you can split this PR.

I really like serializers, I tried the same for filters but the REMAP_FIELDS logic doesn't play well with DRF serializers as you have to remap keys of nested props (e.g. `sector` has to become `sector.id`). It should be doable in a way or another though if we spend more time thinking about it.

@reupen I would keep the pagination logic away from the serializsers, you can use my PaginatedAPIView with the methods that I _borrowed_ from the DRF codebase which should cover all the cases. There's a bit more logic there.

**Comment by reupen** [comment]
@marcofucci I get where you're coming from, but my converse argument would be that it's all in the POST body and should be validated in a consistent way. It could be split to a separate serialiser though.

I'm also not sure I agree with silently ignoring bad values – a negative offset or non-positive limit can't happen by accident, surely? Also if offset <= 10,000 but offset + limit > 10,000 it might be more desirable to limit the number of results instead of returning an error.

**Comment by elcct** [comment]
Are we not going to over-engineer it? It seems like a lot of hassle to just validate a fairly simple request.

**Comment by marcofucci** [comment]
@reupen that's fine, I always try to be consistent even in less desirable cases and I guess we are using DRF pagination elsewhere but I see what you mean as well. 
TBH the DRF code is getting pagination params from GET whilst we are always going to use POST so serializers should feel natural.

**Pull request closed without merging**